src/target/adiv6.c

```diff
@@ -0,0 +1,373 @@
+	const uint16_t ap_reg_base = ADIV5_APnDP | (addr & ADIV6_AP_BANK_MASK);
+
+	uint32_t result = 0;
+	/* Loop through each register location and read it, pulling out only the relevant byte */
+	for (size_t i = 0; i < 4U; ++i) {
+		const uint32_t value = adiv5_dp_read(ap->base.dp, ap_reg_base | (i << 2U));
+		result |= (value & 0xffU) << (i * 8U);
+	}
+	return result;
+}
+
+static uint64_t adiv6_dp_read_pidr(adiv6_access_port_s *const ap)
+{
+	const uint32_t pidr_upper = adiv6_dp_read_id(ap, PIDR4_OFFSET);
+	const uint32_t pidr_lower = adiv6_dp_read_id(ap, PIDR0_OFFSET);
+	return ((uint64_t)pidr_upper << 32U) | (uint64_t)pidr_lower;
+}
+
+static bool adiv6_reset_resources(adiv6_access_port_s *const rom_table)
+{
+	/* Read out power request ID register 0 and check if power control is actually implemented */
+	const uint8_t pridr0 = adiv5_ap_read(&rom_table->base, CORESIGHT_ROM_PRIDR0) & 0x3fU;
+	if ((pridr0 & CORESIGHT_ROM_PRIDR0_VERSION_MASK) != CORESIGHT_ROM_PRIDR0_VERSION_NOT_IMPL)
+		rom_table->base.flags |= ADIV6_DP_FLAGS_HAS_PWRCTRL;
+	/* Now try and perform a debug reset request */
+	if (pridr0 & CORESIGHT_ROM_PRIDR0_HAS_DBG_RESET_REQ) {
+		platform_timeout_s timeout;
+		platform_timeout_set(&timeout, 250);
+
+		adiv5_ap_write(&rom_table->base, CORESIGHT_ROM_DBGRSTRR, CORESIGHT_ROM_DBGRST_REQ);
+		/* While the reset request is in progress */
+		while (adiv5_ap_read(&rom_table->base, CORESIGHT_ROM_DBGRSTRR) & CORESIGHT_ROM_DBGRST_REQ) {
+			/* Check if it's been acknowledge, and if it has, deassert the request */
+			if (adiv5_ap_read(&rom_table->base, CORESIGHT_ROM_DBGRSTAR) & CORESIGHT_ROM_DBGRST_REQ)
+				adiv5_ap_write(&rom_table->base, CORESIGHT_ROM_DBGRSTRR, 0U);
+			/* Check if the reset has timed out */
+			if (platform_timeout_is_expired(&timeout)) {
+				DEBUG_WARN("adiv6: debug reset failed\n");
+				adiv5_ap_write(&rom_table->base, CORESIGHT_ROM_DBGRSTRR, 0U);
+				break;
+			}
+		}
+	}
+	/* Regardless of what happened, extract whether system reset is supported this way */
+	if (pridr0 & CORESIGHT_ROM_PRIDR0_HAS_SYS_RESET_REQ)
+		rom_table->base.flags |= ADIV6_DP_FLAGS_HAS_SYSRESETREQ;
+	return true;
+}
+
+static uint64_t adiv6_read_coresight_rom_entry(
+	adiv6_access_port_s *const rom_table, const uint8_t rom_format, const uint16_t entry_offset)
+{
+	const uint32_t entry_lower = adiv5_ap_read(&rom_table->base, ADIV5_AP_REG(entry_offset));
+	if (rom_format == CORESIGHT_ROM_DEVID_FORMAT_32BIT)
+		return entry_lower;
+	const uint32_t entry_upper = adiv5_ap_read(&rom_table->base, ADIV5_AP_REG(entry_offset + 4U));
+	return ((uint64_t)entry_upper << 32U) | (uint64_t)entry_lower;
+}
+
+static bool adiv6_parse_coresight_v0_rom_table(
+	adiv6_access_port_s *const rom_table, const target_addr64_t base_address, const uint64_t pidr)
+{
+	/* Extract the designer code and part number from the part ID register */
+	const uint16_t designer_code = adiv5_designer_from_pidr(pidr);
```

Comment:
If the functions are generic between adiv6 and adiv5, should the naming of those functions be updated to reflect the fact that they are not specific to the ADI version?

Reply:
We wondered about that but weren't sure about doing so and the resulting churn - but this would definitely be a good idea. We suggest moving those functions and the declarations into adi.c and adi.h as it wouldn't make much sense to leave them in adiv5.{c,h} as a result. If you agree we'll write a short series appended to this to do those moves.

Reply:
Presuming that moving the parts out into a new file is correct, we have made the suggested adjustments.

Reply:
Yeah, moving them to the non versioned files is the correct thing to do. :)